indexing-service/src/main/java/io/druid/indexing/worker/WorkerTaskMonitor.java

```diff
@@ -78,164 +85,354 @@ public WorkerTaskMonitor(
+  private void mainLoop()
   {
-    if (isTaskRunning(task.getId())) {
-      log.warn(
-          "I can't build it. There's something in the way. Got task %s that I am already running...",
-          task.getId()
-      );
-      workerCuratorCoordinator.unannounceTask(task.getId());
-      return;
+    try {
+      while (!Thread.currentThread().isInterrupted()) {
+        final Notice notice = notices.take();
+
+        try {
+          notice.handle();
+        }
+        catch (Exception e) {
+          log.makeAlert(e, "Failed to handle notice")
+             .addData("noticeClass", notice.getClass().getSimpleName())
+             .addData("noticeTaskId", notice.getTaskId())
+             .emit();
+        }
+      }
     }
+    catch (InterruptedException e) {
+      // Exit quietly
+    }
+  }
 
-    log.info("Submitting runnable for task[%s]", task.getId());
+  private void restoreRestorableTasks()
+  {
+    final List<Pair<Task, ListenableFuture<TaskStatus>>> restored = taskRunner.restore();
+    for (Pair<Task, ListenableFuture<TaskStatus>> pair : restored) {
+      addRunningTask(pair.lhs, pair.rhs);
+    }
+  }
 
-    running.add(task);
+  private void cleanupStaleAnnouncements()
+  {
+    // cleanup any old running task announcements which are invalid after restart
+    for (TaskAnnouncement announcement : workerCuratorCoordinator.getAnnouncements()) {
+      if (!running.containsKey(announcement.getTaskStatus().getId()) && announcement.getTaskStatus().isRunnable()) {
+        log.info("Cleaning up stale announcement for task [%s].", announcement.getTaskStatus().getId());
+        workerCuratorCoordinator.updateAnnouncement(
+            TaskAnnouncement.create(
+                announcement.getTaskStatus().getId(),
+                announcement.getTaskResource(),
+                TaskStatus.failure(announcement.getTaskStatus().getId()),
+                TaskLocation.unknown()
+            )
+        );
+      }
+    }
+  }
 
-    exec.submit(
-        new Runnable()
+  private void registerRunListener()
+  {
+    pathChildrenCache.getListenable().addListener(
+        new PathChildrenCacheListener()
         {
           @Override
-          public void run()
+          public void childEvent(CuratorFramework curatorFramework, PathChildrenCacheEvent pathChildrenCacheEvent)
+              throws Exception
           {
-            final long startTime = System.currentTimeMillis();
-
-            TaskStatus taskStatus;
-
-            try {
-              workerCuratorCoordinator.updateAnnouncement(
-                  TaskAnnouncement.create(
-                      task,
-                      TaskStatus.running(task.getId())
-                  )
+            if (pathChildrenCacheEvent.getType().equals(PathChildrenCacheEvent.Type.CHILD_ADDED)) {
+              final Task task = jsonMapper.readValue(
+                  cf.getData().forPath(pathChildrenCacheEvent.getData().getPath()),
```

Comment:
I think such a case would simply have a NoNode exception bubble up past childEvent, skip the rest here, and continue on... which is probably ok.